### PR TITLE
rpcdaemon: improve CBOR decoding

### PR DESCRIPTION
**Pull request by lupin012**

### CODE REVIEW

File: silkworm/silkrpc/ethdb/cbor.cpp [hunk 1]
```diff
@@ -15,6 +15,8 @@
 */
 
 #include "cbor.hpp"
+#include <cbor/cbor.h>
+#include "listener_cbor_log.h"
```
Comment (canepat) [inline silkworm/silkrpc/ethdb/cbor.cpp:19]: Rename `listener_cbor_log.h` as `log_cbor_listener.hpp`

File: silkworm/silkrpc/ethdb/listener_cbor_log.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include <iostream>
+
+#include "listener_cbor_log.h"
```
Comment (canepat) [inline silkworm/silkrpc/ethdb/listener_cbor_log.cpp:19]: This include should be the first one in the file

File: silkworm/silkrpc/ethdb/listener_cbor_log.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (canepat) [inline silkworm/silkrpc/ethdb/listener_cbor_log.cpp:1]: This file should be renamed as `log_cbor_listener.cpp`

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#pragma once
+
+#include <vector>
+
+#include <silkworm/silkrpc/common/log.hpp>
+#include <silkworm/silkrpc/json/types.hpp>
+
+#include <cbor/listener.h>
+
+#define UNUSED(expr) (void)(expr)
```
Comment (canepat) [inline silkworm/silkrpc/ethdb/listener_cbor_log.h:26]: Please avoid using macro except for exceptional situations (e.g. logging macros)

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        void on_null();
+ 
+    private:
+        processing_state state_;
+        int nlogs_;
+        int ntopics_;
+        std::vector<Log>& logs_;
+
+        Log curr_log_;
+        int curr_topic_;
```
Comment (canepat) [inline silkworm/silkrpc/ethdb/listener_cbor_log.h:64]: `curr_topic_` -> `current_topic_`

Comment (canepat) [inline silkworm/silkrpc/ethdb/listener_cbor_log.h:64]: Uninitialised member variable: please initialise it in constructor initialisation list

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        void on_extra_tag(unsigned long long tag) { UNUSED(tag); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(extra_tag)"}; }
+        void on_float32(float value) { UNUSED(value); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(float)"}; }
+        void on_double(double value) { UNUSED(value); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(double)"}; }
+        void on_extra_special(unsigned long long tag) { UNUSED(tag); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(extra_special)"}; }
+        void on_error(const char *error) { UNUSED(error); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(on_error)"}; }
+        void on_special(unsigned int code) { UNUSED(code); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(special)"}; }
+
+        bool is_processing_terminated_successfully();
+
+        std::vector<Log>& get_log()  { return logs_; }
```
Comment (canepat) [inline silkworm/silkrpc/ethdb/listener_cbor_log.h:51]: Not needed

File: silkworm/silkrpc/ethdb/cbor.cpp (hunk 1 above)
Comment (lupin012) [reply silkworm/silkrpc/ethdb/cbor.cpp:19]: moved the new class to cbor.cpp

File: silkworm/silkrpc/ethdb/listener_cbor_log.cpp (hunk 3 above)
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.cpp:1]: file will be removed

File: silkworm/silkrpc/ethdb/listener_cbor_log.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+#include <iostream>
+
+#include "listener_cbor_log.h"
+
+#include <evmc/evmc.hpp>
+#include <silkworm/core/common/util.hpp>
+
+namespace silkrpc {
+
+    void listener_cbor_log::on_bytes(unsigned char *data, int size) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.cpp:26]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.h (hunk 4 above)
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:26]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+           if (++curr_topic_ == ntopics_) {
+              state_ = WAIT_DATA;
+           }
+        } else if (state_ == WAIT_DATA) {
+           curr_log_.data = silkworm::Bytes{data, static_cast<long unsigned int>(size)};
+           logs_.emplace_back(std::move(curr_log_));
+           curr_log_.topics.clear();
+           state_ = WAIT_NFIELDS;
+        } else {
+           throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(on_bytes bad state)"};
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.cpp:43]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        void on_array(int size);
+        void on_null();
+ 
+    private:
+        processing_state state_;
+        int nlogs_;
+        int ntopics_;
+        std::vector<Log>& logs_;
+
+        Log curr_log_;
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:63]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.h (hunk 5 above)
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:64]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.h (hunk 6 above)
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:51]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.cpp (hunk 2 above)
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.cpp:19]: file removed

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+#include <silkworm/silkrpc/common/log.hpp>
+#include <silkworm/silkrpc/json/types.hpp>
+
+#include <cbor/listener.h>
+
+#define UNUSED(expr) (void)(expr)
+
+namespace silkrpc {
+
+    class listener_cbor_log : public cbor::listener {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:30]: ok

[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:30]: ok

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    private:
+        typedef enum { WAIT_NLOGS,  WAIT_NFIELDS, WAIT_ADDRESS, WAIT_NTOPICS, WAIT_TOPICS, WAIT_DATA} processing_state;
+    public:
+        listener_cbor_log (std::vector<Log>& logs): state_(WAIT_NLOGS), nlogs_(0), ntopics_(0), logs_(logs) {}
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:34]: all method are moved into cbor.cpp

File: silkworm/silkrpc/ethdb/listener_cbor_log.h [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        void on_extra_tag(unsigned long long tag) { UNUSED(tag); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(extra_tag)"}; }
+        void on_float32(float value) { UNUSED(value); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(float)"}; }
+        void on_double(double value) { UNUSED(value); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(double)"}; }
+        void on_extra_special(unsigned long long tag) { UNUSED(tag); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(extra_special)"}; }
+        void on_error(const char *error) { UNUSED(error); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(on_error)"}; }
+        void on_special(unsigned int code) { UNUSED(code); throw std::system_error{std::make_error_code(std::errc::invalid_argument), "Log CBOR: unexpected format(special)"}; }
+
+        bool is_processing_terminated_successfully();
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/ethdb/listener_cbor_log.h:49]: ok

**Review by canepat — APPROVED** [review]

**Pull request merged**